### ISSUE TITLE
consolidate existing identifiers

### ISSUE THREAD

**Issue by jspaaks** [citation-file-format/citation-file-format#102]
the following identifiers could be moved to be a value in the ``identifiers`` enum

(root document)

- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L49
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L293 (should become just ``identifiers[].type=doi`` with a ``relationType`` or a ``qualifier`` or something, see https://github.com/citation-file-format/citation-file-format/issues/69)
- arguably ``commit`` is covered by ``identifiers[].url``. In the current schema, the commit is ambiguous because it does not define who is expected to resolve the sha: github, gitlab, bitbucket, software heritage, somebody else?
- ``repository``, ``repository-code``, ``repository-artefact``, ``url``: not sure what to do with these, they can be written down as ``identifier[].type`` with type equals ``url`` 

(references):

- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L293
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L314
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L372
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L431
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L437
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L507
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L533
- https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L606
- ``repository``, ``repository-code``, ``repository-artefact``, ``url``: not sure what to do with these, they can be written down as ``identifier[].type`` with type equals ``url``

**Comment by LourensVeen** [comment]
Interesting! There's a difference between identifiers and locators (e.g. URIs and URLs). Do identifiers have to be resolvable online? What if we're referring to a catalog number of an old book for instance. There may well be scientific collections that have not yet been digitised, but which you could visit to look up the physical book if you have the number of the box it's in. So some of these identifiers may not have online locators. On the other hand, if you have an ISBN, then you can go into any library and see if they have a copy of the book, or go to Amazon or so, so in that case there may be more than one way to resolve. Although you may still no be able to download a copy.

It does seem the case that each identifier lives in a particular context. For some kinds of identifier there's a single global context (DOI, ISBN, ISSN), for others there's one or more specific contexts (Git repository on GitHub, a linked clone on GitLab, or your local clone). Maybe you can also still get the software from the old SVN repo on Sourceforge, and you want to specify two commits, each with their own context.

Given an identifier and its context, we can try to resolve it. For some contexts, there may be only one way to resolve (DOI, NIHM SID, URL), for others there may be multiple (ISBN, Github (SSH or HTTPS, or browse it online)).

Maybe I'm talking nonsense, so let's see what the above turn into using this model:

Current attribute | Identifier | Context | Resolver
------ | ------ | ------ | ------
doi | 10.5281/zenodo.3518011 | (DOI) | curl https://doi.org/<doi\>
collection-doi | 10.5281/zenodo.3258864 | (DOI) | curl https://doi.org/<doi\>
commit | 17c07d6b880e4eec5c54dc50cfb80ebdeffd8e00 | https://github.com/multiscale/muscle3 | git clone
isbn | 978-1441412690 | (ISBN) | your local bookshop, Amazon.com, isbnsearch.org, etc.
issn | 1476-4687 | (ISSN) | university library catalog, ...
nihmsid | NIHMS236863 | (NIH Manuscript Submission system) | https://www.nihms.nih.gov/
pmcid | PMC3531190 | (PubMed Central) | https://www.ncbi.nlm.nih.gov/pmc/
url | https://www.esciencecenter.nl | (The world-wide web) | curl
orcid (?) | 0000-0002-6311-1168 | (ORCID) | https://orcid.org/<orcid\>

The identifier would be given by the user, together with a type attribute which would be mapped to a resolver by any software reading this that supports resolving identifiers. For most of these, the context appears to be implicit, i.e. if you know that something is an ISBN, then the context is the global registry of ISBNs, and there's only one of those. For source code repositories however, there are many of those, so a commit must have a context to be resolvable.

It seems that commit IDs are different because they're not globally unique, as there's not a single global repository containing all source code. Would it make sense to treat them differently from the others? They're maybe more like Docker images (which may be in the global Docker repo, or in a private one) or jar files (I think there are two popular online stores for those).

Also, somewhat related, do we need the possibility of adding multiple identifiers of the same kind? Paper and electronic versions of books for instance have different ISBNs, but may have the same content, and be equally good references. Or should those be considered separate references, each with their own identifier?

**Comment by jspaaks** [comment]
currently, both the root and the ``references`` have an ``identifiers``:
1. https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L62 
1. https://github.com/citation-file-format/citation-file-format/blob/4fa1a91004b3600248bb3f751967e833e7afac63/schema.yaml#L385 

I would advocate that ``commit``s can already be stored using this data structure by including the URL to that specific commit, e.g.:
```yaml
identifiers: 
  - 
    type: url
    value: "https://github.com/citation-file-format/cff-initializer-javascript/tree/646c34dedce6fe2fcc33e2080c70180365c56010"
```

This could be a way of being explicit about the context for ``commit``s

**Comment by jspaaks** [comment]
Regarding multiple identifiers of the same type, one can already have
```yaml
identifiers: 
  - 
    type: doi
    value: "10.5281/zenodo.3539765"
  - 
    type: doi
    value: "10.5281/zenodo.1404735"
```
i.e. the same type of identifier, with one pointing to a versioned doi, the other to the concept doi.

It would be helpful if each one can be annotated, either using a standard vocabulary like datacite's ``relationType`` or maybe even free text?

**Comment by jspaaks** [comment]
orcids I think are different in that they are nested under ``author``

**Comment by LourensVeen** [comment]
True, but ORCIDs are also references to some external entity, so I figured they'd be conceptually similar.

Maybe the types could be extended a bit? So you could have a type doi and a type concept-doi, both of which would contain a doi but with different meaning?

I'm not sure about using the URL to the commit. While you can point a browser to it, this wouldn't allow a program to clone the repo and get the right version I think, but maybe that's not needed. Still, it seems like some information is lost this way. Of course the purpose of this thing is to provide information that allows the reader to find the software, and nothing more, it's not intended to be a general metadata format, so maybe it's good enough.

Would something like a Dockerhub image name/version also fit here?

**Comment by jspaaks** [comment]
I stumbled upon a bunch of identifiers in zenodo's repo:
```
        "ark",
        "arxiv",
        "doi",
        "bibcode",
        "ean8",
        "ean13",
        "eissn",
        "handle",
        "isbn",
        "issn",
        "istc",
        "lissn",
        "lsid",
        "pmid",
        "pmcid",
        "purl",
        "upc",
        "url",
        "urn",
        "orcid",
        "gnd",
        "ads"
```
https://github.com/zenodo/zenodo/blob/22fbff3df385e917cd02a3282ede2d7e6450cb2f/zenodo/modules/deposit/jsonschemas/deposits/records/record-v1.0.0.json#L130-L155

Might be useful later.

**Comment by gjvnq** [comment]
Opinion: identifiers should be just simple strings, preferably URIs.

So instead of:
```yaml
identifiers: 
  - 
    type: doi
    value: "10.5281/zenodo.3539765"
  - 
    type: doi
    value: "10.5281/zenodo.1404735"
  - 
    type: crazy-custom-scheme
    value: "complicatedId"
  -
   type: uuid
   value: "6e8bc430-9c3a-11d9-9669-0800200c9a66"
  -
   type: lex
   value: "eu:council:directive:2010-03-09;2010-19-UE"
```
we would have:
```yaml
identifiers: 
  - "urn:doi:10.5281/zenodo.3539765"
  - "urn:doi:10.5281/zenodo.1404735"
  - "crazy-custom-scheme:complicatedId"
  - "urn:uuid:6e8bc430-9c3a-11d9-9669-0800200c9a66"
  - "urn:lex:eu:council:directive:2010-03-09;2010-19-UE"
```

The list of URNs is available at: https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

**Comment by LourensVeen** [comment]
I agree with that, it's easier to read and write and it looks nicer. Jurriaan's version may be more flexible though, but we may not always need that flexibility.

We're planning to use YAtiML anyway for the CFF Python support, and with that we can have the best of both worlds and actually make it so that you can write them either as a single string as you propose, or in Jurriaan's format (maybe with additional fields), and it would look the same from the software side. Makes it easy (and pretty) for users, while still giving the programmers something simple and consistent to work with.

For future reference, this is described in [the YAtiML docs on parsed classes](https://yatiml.readthedocs.io/en/latest/recipes.html#parsed-classes). I'm [using this trick in yMMSL](https://ymmsl-python.readthedocs.io/en/latest/ymmsl_python.html#simulation-components): in that example, `macro`, `meso` and `micro` are all objects of the same class on the Python side, but a string and two different kinds of mappings (dicts) on the YAML side.

**Comment by jspaaks** [comment]
Hello @gjvnq, and welcome to the project! We appreciate you taking the time to help shape CFF. I quite like the idea of supporting these identifiers in the way you mentioned, and with what @LourensVeen just explained, that seems to be possible. One additional piece of context is that we were considering storing some relation types between identifiers, similar to DataCite's ``relationType`` (see e.g. https://schema.datacite.org/meta/kernel-4.3/doc/DataCite-MetadataKernel_v4.3.pdf). This came out of a discussion of differentiating between the conceptDOI and the versioned DOI in a given CITATION.cff. Although I myself am not convinced that this is the right way to go for CFF; I think this may perhaps be more suited for [the CodeMeta project](https://codemeta.github.io/).

**Comment by sdruskat** [comment]
As of v1.2.0, we also have a (human-readable) description for identifiers, which may help users (but not machines) to tell, e.g., different DOIs apart.
